flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/TestFlinkUpsert.java

```diff
@@ -0,0 +1,302 @@
+          tEnv = StreamTableEnvironment.create(env, settingsBuilder.build());
+        } else {
+          settingsBuilder.inBatchMode();
+          tEnv = TableEnvironment.create(settingsBuilder.build());
+        }
+      }
+    }
+    return tEnv;
+  }
+
+  @Override
+  @Before
+  public void before() {
+    super.before();
+    sql("CREATE DATABASE IF NOT EXISTS %s", flinkDatabase);
+    sql("USE CATALOG %s", catalogName);
+    sql("USE %s", DATABASE);
+  }
+
+  @Override
+  @After
+  public void clean() {
+    sql("DROP DATABASE IF EXISTS %s", flinkDatabase);
+    BoundedTableFactory.clearDataSets();
+    super.clean();
+  }
+
+  @Test
+  public void testUpsertAndQuery() {
+    String tableName = "test_upsert_query";
+
+    sql("CREATE TABLE %s(id INT NOT NULL, province STRING NOT NULL, dt DATE, PRIMARY KEY(id,province) NOT ENFORCED) " +
+            "PARTITIONED BY (province) WITH %s",
+        tableName, toWithClause(tableUpsertProps));
+
+    try {
+      sql("INSERT INTO %s VALUES " +
+          "(1, 'a', TO_DATE('2022-03-01'))," +
+          "(2, 'b', TO_DATE('2022-03-01'))," +
+          "(1, 'b', TO_DATE('2022-03-01'))",
+          tableName);
+
+      sql("INSERT INTO %s VALUES " +
+          "(4, 'a', TO_DATE('2022-03-02'))," +
+          "(5, 'b', TO_DATE('2022-03-02'))," +
+          "(1, 'b', TO_DATE('2022-03-02'))",
+          tableName);
+
+      List<Row> result = sql("SELECT * FROM %s WHERE dt < '2022-03-02'", tableName);
+
+      Assert.assertEquals("result should have 2 rows!", 2, result.size());
+
+      result = sql("SELECT * FROM %s WHERE dt < '2022-03-03'", tableName);
+
+      Assert.assertEquals("result should have 5 rows!", 5, result.size());
+
+      sql("DROP TABLE IF EXISTS %s.%s", flinkDatabase, tableName);
+    } finally {
+      sql("DROP TABLE IF EXISTS %s.%s", flinkDatabase, tableName);
+    }
+  }
+
+  @Test
+  public void testPrimaryKeyKEqualToPartitionKey() {
+    // This is an SQL based reproduction of TestFlinkIcebergSinkV2#testUpsertOnDataKey
+    String tableName = "upsert_on_data_key";
+    try {
+      sql("CREATE TABLE %s(id INT NOT NULL, data STRING NOT NULL, PRIMARY KEY(data) NOT ENFORCED) " +
+              "PARTITIONED BY (data) WITH %s",
+          tableName, toWithClause(tableUpsertProps));
+
+      sql("INSERT INTO %s VALUES " +
+          "(1, 'aaa')," +
+          "(2, 'aaa')," +
+          "(3, 'bbb')",
+          tableName);
+
+      List<Row> result = sql("SELECT * FROM %s", tableName);
+
+      Assert.assertEquals("result should have 2 rows!", 2, result.size());
+      Assert.assertEquals("result should have the correct rows",
+          Sets.newHashSet(Row.of(2, "aaa"), Row.of(3, "bbb")), Sets.newHashSet(result));
+
+      sql("INSERT INTO %s VALUES " +
+          "(4, 'aaa')," +
+          "(5, 'bbb')",
+          tableName);
+
+      result = sql("SELECT * FROM %s", tableName);
+
+      // Assert.assertEquals("result should have 2 rows!", 2, result.size());
```

Comment:
Does this need to be here?

Reply:
Nope. This was left over from testing, so that we could see the rows if they were failing (and not just the count).

Uncommenting these, though I'd also be fine with removing them entirely.

Reply:
Actually, since we check the rows immediately afterwards, this check is unnecessary. I'm just going to remove it.

Reply:
Removed the checks on row count for the tests that explicitly check the values of the rows right after.

They arguably have _some_ small benefit, because we do a hash set comparison on the rows so duplicates might not properly be accounted for.

But I think they're fine to skip. I've not encountered an occurrence of running the tests where there was a duplicate row in the results - which is all that these `result.size()` checks are testing.